libsql/src/sync.rs

```diff
@@ -1,18 +1,149 @@
 const DEFAULT_MAX_RETRIES: usize = 5;
+
+use bytes::Bytes;
+use http::{HeaderValue, Request, Uri};
+use tokio::sync::Mutex;
+
+use crate::{util::ConnectorService, Result};
+
 pub struct SyncContext {
-    pub sync_url: String,
-    pub auth_token: Option<String>,
-    pub max_retries: usize,
-    pub durable_frame_num: u32,
+    sync_url: String,
+    auth_token: Option<String>,
+    max_retries: usize,
+    durable_frame_num: u32,
+    db_path: String,
+    max_frame_no: u32,
+
+    client: hyper::Client<ConnectorService, hyper::Body>,
 }
 
 impl SyncContext {
-    pub fn new(sync_url: String, auth_token: Option<String>) -> Self {
-        Self {
+    pub async fn new(
+        sync_url: String,
+        auth_token: Option<String>,
+        db_path: impl Into<String>,
+        connector: ConnectorService,
+    ) -> Self {
+        let mut ctx = Self {
             sync_url,
             auth_token,
             durable_frame_num: 0,
             max_retries: DEFAULT_MAX_RETRIES,
+            db_path: db_path.into(),
+            max_frame_no: 0,
+            client: hyper::Client::builder().build(connector),
+        };
+
+        ctx.read_and_update_metadata().await.unwrap();
+
+        ctx
+    }
+
+    pub(crate) async fn send_frame(
+        &mut self,
+        frame: Bytes,
+        generation: u32,
+        frame_no: u32,
+    ) -> Result<u32> {
+        let url = format!(
+            "{}/sync/{}/{}/{}",
+            self.sync_url,
+            generation,
+            frame_no,
+            frame_no + 1
+        );
+
+        let maybe_auth_header = if let Some(auth_token) = &self.auth_token {
+            Some(HeaderValue::from_str(&format!("Bearer {}", auth_token)).unwrap())
+        } else {
+            None
+        };
+
+        let mut attempts = 0;
+
+        loop {
+            let mut req = Request::post(url.clone());
+
+            if let Some(auth_header) = &maybe_auth_header {
+                req.headers_mut()
+                    .unwrap()
+                    .insert("Authorization", auth_header.clone());
+            }
+
+            let req = req.body(frame.clone().into()).unwrap();
+
+            let res = self.client.request(req).await.unwrap();
+
+            if res.status().is_success() {
+                let body = hyper::body::to_bytes(res.into_body()).await.unwrap();
+
+                let resp = serde_json::from_slice::<serde_json::Value>(&body[..]).unwrap();
+
+                let max_frame_no = resp.get("max_frame_no").unwrap().as_u64().unwrap() as u32;
+
+                // Update our best known max_frame_no from the server and write it to disk.
+                self.set_max_frame_no(max_frame_no).await.unwrap();
+
+                return Ok(max_frame_no);
+            } else if res.status().is_server_error() || attempts < self.max_retries {
```

Comment:
@penberg small change compared to what you had, I just retry on server failures rather than all failures like auth_token being invalid.